File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int lsf = Integer.MAX_VALUE; //leastSOFar
        int overallProfit = 0;
        int pist = 0; //profit if sold today
        
        for (int i = 0; i < prices.length; i++){
            if(prices[i] < lsf)
                lsf = prices[i];
            pist = prices[i] - lsf;
            
            if(overallProfit < pist)
                overallProfit = pist;
        }
        
        return overallProfit;
    }
}

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        profit = windowStart = 0
        min_price = float('inf')
        for windowEnd in range(len(prices)):
            if prices[windowEnd] < min_price:
                min_price = prices[windowEnd]
            elif (prices[windowEnd] - min_price > profit):
                profit = prices[windowEnd] - min_price
                
        return profit